src/deprecated.jl

```diff
@@ -1 +1,53 @@
 Base.@deprecate_binding NO_FIELDS NoTangent()
+
+const EXTERN_DEPRECATION = "`extern` is deprecated, use `unthunk` or `backing` instead, " *
+    "depending on the use case."
+
+"""
+    extern(x)
+
+Makes a best effort attempt to convert a differential into a primal value.
+This is not always a well-defined operation.
+For two reasons:
+ - It may not be possible to determine the primal type for a given differential.
+ For example, `Zero` is a valid differential for any primal.
+ - The primal type might not be a vector space, thus might not be a valid differential type.
+ For example, if the primal type is `DateTime`, it's not a valid differential type as two
+ `DateTime` can not be added (fun fact: `Milisecond` is a differential for `DateTime`).
+
+Where it is defined the operation of `extern` for a primal type `P` should be
+`extern(x) = zero(P) + x`.
+
+!!! note
+    Because of its limitations, `extern` should only really be used for testing.
+    It can be useful, if you know what you are getting out, as it recursively removes
+    thunks, and otherwise makes outputs more consistent with finite differencing.
+
+    The more useful action in general is to call `+`, or in the case of a [`Thunk`](@ref)
+    to call [`unthunk`](@ref).
+
+!!! warning
+    `extern` may return an alias (not necessarily a copy) to data
+    wrapped by `x`, such that mutating `extern(x)` might mutate `x` itself.
+"""
+@inline function extern(x)
+    Base.depwarn(EXTERN_DEPRECATION, :extern)
```

Comment:
I thought I asked this before, but if so github has eaten the reply.
(Probably i forget to ask)

Why are we using `Base.depwarn` rather than `@deprecate`?

Reply:
Oh, I think it's because we are not providing a definite alternative. I.e. it isn't always just `extern` -> `unthunk`, it could be using `backing` as well

Reply:
In general we are not.
But for any specific case they can make the change to put the RHS of the deprecation in the place of the `extern`.
which they might have to do recursively til it stops but still.
Though perhaps that is worse.